src/common/service/NetRemoteDataStreamingReactors.cxx

```diff
@@ -235,3 +235,121 @@ DataStreamWriter::HandleFailure(const std::string& errorMessage)
+
+    if (isOk) {
+        m_numberOfDataBlocksReceived++;
+        m_status.set_code(DataStreamOperationStatusCode::DataStreamOperationStatusCodeSucceeded);
+        m_status.set_message("Data read successful");
+        StartRead(&m_readData);
+    } else {
+        // A false "isOk" value could either mean a failed RPC or that no more data is available.
+        // Unfortunately, there is no clear way to tell which situation occurred.
+        bool readsDoneExpected{ false };
+        if (m_readsDone.compare_exchange_strong(readsDoneExpected, true, std::memory_order_relaxed, std::memory_order_relaxed)) {
+            Finish(grpc::Status::OK);
+        }
+    }
+}
+
+void
+DataStreamReaderWriter::OnWriteDone(bool isOk)
+{
+    const FunctionTracer traceMe{};
+
+    // Client may have canceled the RPC, so check for cancelation to prevent writing more data
+    // when we shouldn't.
+    if (m_isCanceled.load(std::memory_order_relaxed)) {
+        LOGD << "RPC canceled, returning early";
+        return;
+    }
+
+    // Check for a failed status code from HandleWriteFailure since that invoked a final write, thus causing this callback to be invoked.
+    if (m_status.code() == DataStreamOperationStatusCode::DataStreamOperationStatusCodeFailed) {
+        Finish(::grpc::Status::OK);
+        return;
+    }
+
+    // Continue writing if previous write was successful, otherwise handle the failure.
+    if (isOk) {
+        m_status.set_code(DataStreamOperationStatusCode::DataStreamOperationStatusCodeSucceeded);
+        m_status.set_message("Data write successful");
+        NextWrite();
+    } else {
+        // If OnReadDone() failed (due to no more data sent from the client), then Finish() was called and this write will fail.
+        // In that case, do not call HandleFailure() because that triggers another write. Otherwise, this is a true write failure.
+        if (!m_readsDone.load(std::memory_order_relaxed)) {
+            HandleFailure("Data write failed");
+        }
+    }
+}
+
+void
+DataStreamReaderWriter::OnCancel()
+{
+    const FunctionTracer traceMe{};
+
+    // The RPC is canceled by the client, so call Finish to complete it from the server perspective.
+    bool isCanceledExpected{ false };
+    if (m_isCanceled.compare_exchange_strong(isCanceledExpected, true, std::memory_order_relaxed, std::memory_order_relaxed)) {
+        Finish(grpc::Status::CANCELLED);
+    }
+}
+
+void
+DataStreamReaderWriter::OnDone()
+{
+    const FunctionTracer traceMe{};
+    delete this;
+}
+
+void
+DataStreamReaderWriter::NextWrite()
+{
+    const FunctionTracer traceMe{};
+
+    // Client may have canceled the RPC, so check for cancelation to prevent writing more data
+    // when we shouldn't.
+    if (m_isCanceled.load(std::memory_order_relaxed)) {
+        LOGD << "RPC canceled, aborting write";
+        return;
+    }
+
+    // Create data to write to the client.
+    const auto data = m_dataGenerator.GenerateRandomData();
```

Comment:
Are there plans to allow configuration of the block size?

Reply:
Not currently. Because this API doesn't have a "request"/"response" like regular RPCs and instead has a "stream" for both directions, there is no intuitive way to allow the client to specify a block size that it wants the server to use. I think that if there was a scenario that required that, then the DataStreamDownload API should be used instead

Reply:
If this is a requirement (I think it is), there are a few ways we could consider achieving this:

1. Since all fields are optional in proto3 format, the first input message from the client could populate a field that specifies properties about the requested stream from the server. This field is only inspected on the first message received and otherwise not even checked for.
2. Streaming could be stateful, where a request is first made to create a stream, and then that stream has to be explicitly started with the usual API. This way, both the client and server can negotiate stream properties. It could look something like this:

```proto
enum DataStreamDirection
{
    Upload,
    Download,
    Both,
};
enum DataStreamCreateStatusCode 
{
    Unknown = 0,
    Succeeded = 1,
    Failed = 2,
    // Others as needed
};
message DataStreamCreateRequest
{
    DataStreamDirection Direction = 1;
    DataStreamProperties PropertiesUpload = 2; // present when Direction == Upload and Both
    DataStreamProperties PropertiesDownload = 2; // present when Direction == Download and Both
};
message DataStreamCreateResult
{
    DataStreamCreateStatusCode Status = 0;
    string StreamId = 1; // Valid/present only when  Status == Succeeded
};
message DataStreamDownloadRequest
{
    string StreamId = 0;
};
rpc DataStreamCreate (DataStreamCreateRequest request, DataStreamCreateResult result);
```

Reply:
Tracking this in #206 and will look into it later. Your first suggestion is definitely simpler and stays closer to the original design, but your second suggestion is cleaner and simple to understand, although requires a bit more work and changes the design a bit.

Reply:
Thanks! Both require logic on the server side to check for the presence of a field, and while intuitive for the client, having conditional logic like that feels wrong. Anyway, we can drill into this as part of #206 after we determine if this is really a requirement of not.